src/test/ui/print_type_sizes/padding.stdout

```diff
@@ -1,23 +1,21 @@
-print-type-size type: `E1`: 12 bytes, alignment: 4 bytes
-print-type-size     discriminant: 1 bytes
-print-type-size     variant `B`: 11 bytes
-print-type-size         padding: 3 bytes
-print-type-size         field `.0`: 8 bytes, alignment: 4 bytes
-print-type-size     variant `A`: 7 bytes
-print-type-size         field `.1`: 1 bytes
+print-type-size type: `E1`: 8 bytes, alignment: 4 bytes
+print-type-size     variant `A`: 8 bytes
+print-type-size         padding: 1 bytes
+print-type-size         field `.1`: 1 bytes, alignment: 1 bytes
 print-type-size         padding: 2 bytes
 print-type-size         field `.0`: 4 bytes, alignment: 4 bytes
-print-type-size type: `E2`: 12 bytes, alignment: 4 bytes
-print-type-size     discriminant: 1 bytes
-print-type-size     variant `B`: 11 bytes
-print-type-size         padding: 3 bytes
-print-type-size         field `.0`: 8 bytes, alignment: 4 bytes
-print-type-size     variant `A`: 7 bytes
-print-type-size         field `.0`: 1 bytes
+print-type-size     variant `B`: 8 bytes
+print-type-size         field `.0`: 8 bytes
+print-type-size type: `E2`: 8 bytes, alignment: 4 bytes
+print-type-size     variant `A`: 8 bytes
+print-type-size         padding: 1 bytes
+print-type-size         field `.0`: 1 bytes, alignment: 1 bytes
 print-type-size         padding: 2 bytes
 print-type-size         field `.1`: 4 bytes, alignment: 4 bytes
+print-type-size     variant `B`: 8 bytes
+print-type-size         field `.0`: 8 bytes
 print-type-size type: `S`: 8 bytes, alignment: 4 bytes
-print-type-size     field `.g`: 4 bytes
-print-type-size     field `.a`: 1 bytes
 print-type-size     field `.b`: 1 bytes
-print-type-size     end padding: 2 bytes
+print-type-size     field `.a`: 1 bytes
+print-type-size     padding: 2 bytes
```

Comment:
Would a padding in the middle be better than a start padding or end padding?

Reply:
Hmm, it shouldn't be reordering this struct, since the largest niche is already at the first index.

Reply:
> Would a padding in the middle be better than a start padding or end padding?

Yes, if it allow to move the niche at the beginning of the struct.
This allow to use the niche optimization more often if the struct is used in enum. And this is the case here as `E1` and `E2` have reduced in size.

(Note that the niche unfortunately cannot use the padding, only types which have invalid value (so `bool` in this case). see discussion https://github.com/rust-lang/rust/issues/70230 )

Reply:
Pushed a change to avoid reordering `a` and `b` WRT each other here. Apparently `max_by_key` picks the last maximum element if there are multiple, unlike `min_by_key` which picks the first.

And yeah moving niche to the beginning maximizes space after the niche for other variants ^

Reply:
I saw the change but the padding is still in the middle, wouldn't it be faster to have the data aligned together where the padding be on the start or the end?

Reply:
That's a good point: Maybe one should do a test with only the change that re-order the fields (23d815c679e777c57632a914ea600b9b66a46071 and 059b19a761c6ad7890a9b0ffbc85698966cc1528) 
to see if the this could be the cause of the performance degradation.

Reply:
Can bors even run multiple times for each commit? I thought a separate pull request is needed?

Reply:
Yeah, once we get perf results for the full change it makes sense to try just the reordering to narrow it down.